src/main/java/at/ac/tuwien/kr/alpha/grounder/Substitution.java

```diff
@@ -68,29 +75,98 @@ public static Substitution unify(Literal literal, Instance instance, Substitutio
+				// All terms unify but there was no need to assign a new variable, return the input substitution.
+				return partialSubstitution;
 			}
-			return null;
+			return returnSubstitution;
 		}
-		return substitution;
+
+		boolean unifyTerms(Term termNonGround, Term termGround, Substitution partialSubstitution) {
+			if (termNonGround == termGround) {
+				// Both terms are either the same constant or the same variable term
+				return true;
+			} else if (termNonGround instanceof ConstantTerm) {
+				// Since right term is ground, both terms differ
+				return false;
+			} else if (termNonGround instanceof VariableTerm) {
+				VariableTerm variableTerm = (VariableTerm) termNonGround;
+				// Left term is variable, bind it to the right term. Use original substitution if it has
+				// not been cloned yet.
+				Term bound = (returnSubstitution == null ? partialSubstitution : returnSubstitution).eval(variableTerm);
+				if (bound != null) {
+					// Variable is already bound, return true if binding is the same as the current ground term.
+					return termGround == bound;
+				}
+				// Record new variable binding.
+				if (returnSubstitution == null) {
+					// Clone substitution if it was not yet.
+					returnSubstitution = new Substitution(partialSubstitution);
+				}
+				returnSubstitution.put(variableTerm, termGround);
+				return true;
+			} else if (termNonGround instanceof FunctionTerm && termGround instanceof FunctionTerm) {
+				// Both terms are function terms
+				FunctionTerm ftNonGround = (FunctionTerm) termNonGround;
+				FunctionTerm ftGround = (FunctionTerm) termGround;
+
+				if (!(ftNonGround.getSymbol().equals(ftGround.getSymbol()))) {
+					return false;
+				}
+				if (ftNonGround.getTerms().size() != ftGround.getTerms().size()) {
+					return false;
+				}
+
+				// Iterate over all subterms of both function terms
+				for (int i = 0; i < ftNonGround.getTerms().size(); i++) {
+					if (!unifyTerms(ftNonGround.getTerms().get(i), ftGround.getTerms().get(i), partialSubstitution)) {
+						return false;
+					}
+				}
+
+				return true;
+			}
+			return false;
+		}
+	}
+
+	/**
+	 * Computes the unifier of the (nonground) atom, a partial substitution for the atom, and a ground instance.
+	 * If unification succeeds the unifying substitution is returned. The partial substitution is left unchanged.
+	 *
+	 * @param atom         the (potentially nonground) atom to unify.
+	 * @param instance     the ground instance to unify with.
+	 * @param substitution the (partial) substitution for the atom. This is left unchanged in all cases.
+	 * @return null if the unification fails, otherwise it is a unifying substitution. If the parameter substitution
+	 * 	already is a unifier, it is returned. If the unifying substitution is an extension of the input
+	 * 	substitution, a new substitution will be returned.
+	 */
+	public static Substitution unify(Atom atom, Instance instance, Substitution substitution) {
```

Comment:
More of a general question, and maybe out of the scope of this review... We also have classes `Unifier` and `Unification`, with `Uninfication` implementing utility functions looking very similar to some of the methods in `Substitution`. In order to avoid duplicating code, or writing several very similar versions, would it make sense to use the (accordingly adapted) methods from `Unification` here?

Reply:
There is a slight difference between the two (and maybe we should adapt our naming): the one in `Substitution` is assuming that variables only ever occur on the left-hand side (i.e., the second parameter is always ground). The version in `Unification` really computes the unifier between two non-ground `Atom`s. Maybe we should rename the methods in `Substitution` to `instantiate()` instead of `unify()`.

Reply:
I decided to rename `Substitution::unify` to `Substitution::specialize` as it specializes the given partial substitution to also unify the given atom and the instance.